language/move-vm/runtime/src/interpreter.rs

```diff
@@ -262,18 +265,45 @@ impl Interpreter {
     ///
     /// Native functions do not push a frame at the moment and as such errors from a native
     /// function are incorrectly attributed to the caller.
-    fn make_call_frame(&mut self, func: Arc<Function>, ty_args: Vec<Type>) -> VMResult<Frame> {
+    fn make_call_frame(
+        &mut self,
+        func: Arc<Function>,
+        ty_args: Vec<Type>,
+        loader: &Loader,
+    ) -> VMResult<Frame> {
+        let resolver = func.get_resolver(loader);
         let mut locals = Locals::new(func.local_count());
         let arg_count = func.arg_count();
         for i in 0..arg_count {
+            let v = self.operand_stack.pop().map_err(|e| self.set_location(e))?;
+            if self.strict {
+                let expected_ty = resolver
+                    .resolve_signature_token(&func.local_types()[arg_count - i - 1])
+                    .and_then(|ty| ty.subst(&ty_args))
+                    .map_err(|e| match func.module_id() {
+                        Some(id) => e
+                            .at_code_offset(func.index(), 0)
+                            .finish(Location::Module(id.clone())),
+                        None => {
+                            let err =
+                                PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                                    .with_message(
+                                        "Unexpected native function not located in a module"
+                                            .to_owned(),
+                                    );
+                            self.set_location(err)
+                        }
+                    })?;
+
+                v.check_type(&expected_ty)
```

Comment:
This is also rather expensive if we have to do this for every call. I would expect we have some cached type hash info for the expected arguments of this function.